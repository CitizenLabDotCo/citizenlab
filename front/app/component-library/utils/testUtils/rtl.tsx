/* eslint-disable no-restricted-imports */
import React, { FC } from 'react';

import { render, RenderOptions } from '@testing-library/react';
import { HelmetProvider } from 'react-helmet-async';
import { ThemeProvider } from 'styled-components';

import GlobalStyle from '../../global-styles';
import { getTheme } from '../styleUtils';

const AllTheProviders: FC = ({ children }) => {
  return (
    <HelmetProvider>
      <ThemeProvider theme={getTheme(null)}>
        <GlobalStyle />
        {children}
      </ThemeProvider>
    </HelmetProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
