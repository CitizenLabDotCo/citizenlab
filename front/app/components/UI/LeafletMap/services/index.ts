import { init, addTileLayer, changeView } from './setup';

import { addLayersControl, removeLayersControl } from './controls';
import { addLayers, removeLayers, removeLayer } from './layers';
import {
  addMarkersToMap,
  addMarkerClusterGroup,
  getMarkerIcon,
} from './markers';

export default {
  init,
  addTileLayer,
  changeView,
  addLayersControl,
  removeLayersControl,
  addLayers,
  removeLayers,
  removeLayer,
  addMarkersToMap,
  addMarkerClusterGroup,
  getMarkerIcon,
};
