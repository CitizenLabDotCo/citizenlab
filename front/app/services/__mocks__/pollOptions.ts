export const mockOption = (id, titleEn) => ({
  id,
  attributes: {
    title_multiloc: {
      en: titleEn,
    },
  },
});
