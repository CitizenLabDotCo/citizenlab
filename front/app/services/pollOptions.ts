import { API_PATH } from 'containers/App/constants';
import streams from 'utils/streams';
import { Multiloc } from 'typings';

export function updatePollOption(optionId: string, titleMultiloc: Multiloc) {
  return streams.update(`${API_PATH}/poll_options/${optionId}`, optionId, {
    title_multiloc: titleMultiloc,
  });
}
