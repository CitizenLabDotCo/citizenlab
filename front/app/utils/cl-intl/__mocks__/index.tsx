import React from 'react';
const Intl = jest.requireActual('react-intl');

// Initialise the real provider so that we don't
// need to reimplement any internals
const defaultProps = {
  locale: 'en',
  defaultLocale: 'en',
};
const intlProvider = new Intl.IntlProvider(defaultProps, {});

// The exact same `intl` object the real code receives ;-)
const { intl } = intlProvider.getChildContext();

const formatMessageReplacement = (
  messageDescriptor: ReactIntl.FormattedMessage.MessageDescriptor,
  values?: { [key: string]: string | number | boolean | Date } | undefined
) => {
  return intl.formatMessage(messageDescriptor, {
    tenantName: 'The Test',
    orgName: 'Test Town',
    orgType: 'testing',
    ...(values || {}),
  });
};

const intlReplacement = {
  ...intl,
  formatMessage: formatMessageReplacement,
};

const injectIntlReplacement = (Node) => {
  const renderWrapped = (props) => <Node {...props} intl={intl} />;
  renderWrapped.displayName = Node.displayName || Node.name || 'Component';
  return renderWrapped;
};

Intl.injectIntl = injectIntlReplacement;

// Override components by implementing the real components
//  providing them the context they need in order to function
const {
  IntlProvider,
  FormattedDate,
  FormattedTime,
  FormattedRelative,
  FormattedNumber,
  FormattedPlural,
  FormattedMessage,
  FormattedHTMLMessage,
} = Intl;

Intl.FormattedDate = (props) => (
  <IntlProvider {...defaultProps}>
    <FormattedDate {...props} />
  </IntlProvider>
);
Intl.FormattedTime = (props) => (
  <IntlProvider {...defaultProps}>
    <FormattedTime {...props} />
  </IntlProvider>
);
Intl.FormattedRelative = (props) => (
  <IntlProvider {...defaultProps}>
    <FormattedRelative {...props} />
  </IntlProvider>
);
Intl.FormattedNumber = (props) => (
  <IntlProvider {...defaultProps}>
    <FormattedNumber {...props} />
  </IntlProvider>
);
Intl.FormattedPlural = (props) => (
  <IntlProvider {...defaultProps}>
    <FormattedPlural {...props} />
  </IntlProvider>
);
Intl.FormattedMessage = (props) => (
  <IntlProvider {...defaultProps}>
    <FormattedMessage {...props} />
  </IntlProvider>
);
Intl.FormattedHTMLMessage = (props) => (
  <IntlProvider {...defaultProps}>
    <FormattedHTMLMessage {...props} />
  </IntlProvider>
);

// Set displayName so that snapshots don't use "Uknown" as component name
Intl.FormattedDate.displayName = 'FormattedDate';
Intl.FormattedTime.displayName = 'FormattedTime';
Intl.FormattedRelative.displayName = 'FormattedRelative';
Intl.FormattedNumber.displayName = 'FormattedNumber';
Intl.FormattedPlural.displayName = 'FormattedPlural';
Intl.FormattedMessage.displayName = 'FormattedMessage';
Intl.FormattedHTMLMessage.displayName = 'FormattedHTMLMessage';

// Special hook for tests, real package does not export this
Intl.intl = intlReplacement;

module.exports = Intl;
