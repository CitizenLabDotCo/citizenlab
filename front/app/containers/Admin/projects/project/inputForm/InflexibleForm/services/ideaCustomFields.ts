import { API_PATH } from 'containers/App/constants';
import streams, { IStreamParams } from 'utils/streams';
import { Multiloc } from 'typings';
import { CustomFieldCodes } from 'services/ideaCustomFieldsSchemas';

export type Visibility = 'admins' | 'public';

export interface IIdeaCustomFieldData {
  id: string;
  type: 'custom_field';
  attributes: {
    key: string;
    input_type:
      | 'text'
      | 'number'
      | 'multiline_text'
      | 'html'
      | 'text_multiloc'
      | 'multiline_text_multiloc'
      | 'html_multiloc'
      | 'select'
      | 'multiselect'
      | 'checkbox'
      | 'date'
      | 'files'
      | 'image_files';
    title_multiloc: Multiloc;
    description_multiloc: Multiloc;
    required: boolean;
    ordering: null;
    enabled: boolean;
    code: CustomFieldCodes;
    created_at: null;
    updated_at: null;
    visible_to: Visibility;
  };
}

export interface IIdeaCustomField {
  data: IIdeaCustomFieldData;
}

export interface IIdeaCustomFields {
  data: IIdeaCustomFieldData[];
}

export interface IUpdatedIdeaCustomFieldProperties {
  description_multiloc?: Multiloc;
  enabled?: boolean;
  required?: boolean;
  visible_to?: Visibility;
}

export function ideaCustomFieldsStream(
  projectId: string,
  streamParams: IStreamParams | null = null
) {
  const apiEndpoint = `${API_PATH}/admin/projects/${projectId}/custom_fields`;
  return streams.get<IIdeaCustomFields>({ apiEndpoint, ...streamParams });
}

export async function updateIdeaCustomField(
  projectId: string,
  ideaCustomFieldId: string,
  code: string | undefined,
  object: IUpdatedIdeaCustomFieldProperties
) {
  const apiEndpoint = code
    ? `${API_PATH}/admin/projects/${projectId}/custom_fields/by_code/${code}`
    : `${API_PATH}/admin/projects/${projectId}/custom_fields/update/${ideaCustomFieldId}`;
  const updateObject = { custom_field: object };
  return streams.update<IIdeaCustomField>(
    apiEndpoint,
    ideaCustomFieldId,
    updateObject
  );
}

export async function refetchCustomFields(projectId) {
  const apiEndpoint = `${API_PATH}/admin/projects/${projectId}/custom_fields`;
  streams.fetchAllWith({ apiEndpoint: [apiEndpoint] });
}
