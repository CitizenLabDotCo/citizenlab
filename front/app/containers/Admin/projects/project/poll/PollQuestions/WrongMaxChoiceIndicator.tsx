import React from 'react';
import { isNilOrError } from 'utils/helperUtils';
import messages from '../messages';
import { FormattedMessage } from 'utils/cl-intl';
import { TextCell } from 'components/admin/ResourceList';
import styled from 'styled-components';
import usePollOptions from 'hooks/usePollOptions';
import { colors, IconTooltip } from '@citizenlab/cl2-component-library';

const StyledIconTooltip = styled(IconTooltip)`
  margin-right: 5px;
`;

const Indicator = styled(TextCell)<{ isWarning?: boolean }>`
  display: flex;
  color: ${({ isWarning }) => (isWarning ? colors.orange : colors.error)};
  margin-right: 15px;
`;

interface Props {
  questionId: string;
  maxAnswers: number | null;
}

const WrongMaxChoiceIndicator = ({ maxAnswers, questionId }: Props) => {
  const options = usePollOptions(questionId);

  if (isNilOrError(options) || typeof maxAnswers !== 'number') {
    return null;
  }

  return options.length < maxAnswers ? (
    <Indicator isWarning data-testid="wrongMaxChoiceIndicator">
      <StyledIconTooltip
        content={<FormattedMessage {...messages.maxOverTheMaxTooltip} />}
      />
      <FormattedMessage {...messages.wrongMax} />
    </Indicator>
  ) : null;
};

export default WrongMaxChoiceIndicator;
