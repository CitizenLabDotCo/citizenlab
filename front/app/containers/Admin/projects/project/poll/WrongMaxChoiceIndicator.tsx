import React from 'react';
import GetPollOptions, {
  GetPollOptionsChildProps,
} from 'resources/GetPollOptions';
import { adopt } from 'react-adopt';
import { isNilOrError } from 'utils/helperUtils';
import messages from './messages';
import { FormattedMessage } from 'utils/cl-intl';
import { StyledIconTooltip, Indicator } from './WrongOptionsIndicator';
import styled from 'styled-components';

const StyledIndicator = styled(Indicator)`
  margin-right: 15px;
`;

interface InputProps {
  questionId: string;
  maxAnswers: number | null;
}

interface DataProps {
  options: GetPollOptionsChildProps;
}

interface Props extends InputProps, DataProps {}

export const WrongMaxChoiceIndicator = ({ options, maxAnswers }: Props) =>
  typeof maxAnswers === 'number' && maxAnswers < 2 ? (
    <StyledIndicator>
      <StyledIconTooltip
        content={<FormattedMessage {...messages.maxUnderTheMinTooltip} />}
      />
      <FormattedMessage {...messages.wrongMax} />
    </StyledIndicator>
  ) : !isNilOrError(options) && maxAnswers && options.length < maxAnswers ? (
    <StyledIndicator isWarning>
      <StyledIconTooltip
        content={<FormattedMessage {...messages.maxOverTheMaxTooltip} />}
      />
      <FormattedMessage {...messages.wrongMax} />
    </StyledIndicator>
  ) : null;

const Data = adopt<DataProps, InputProps>({
  options: ({ questionId, render }) => (
    <GetPollOptions questionId={questionId}>{render}</GetPollOptions>
  ),
});

export default (inputProps: InputProps) => (
  <Data {...inputProps}>
    {(dataprops) => <WrongMaxChoiceIndicator {...inputProps} {...dataprops} />}
  </Data>
);
