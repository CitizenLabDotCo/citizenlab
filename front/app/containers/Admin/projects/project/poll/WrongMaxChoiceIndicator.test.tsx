import React from 'react';
import WrongMaxChoiceIndicator from './WrongMaxChoiceIndicator';
import { mockOption } from 'services/__mocks__/pollOptions';
import { render, screen } from 'utils/testUtils/rtl';
import { IPollOptionData } from 'services/pollOptions';

const mockPollOptions: IPollOptionData[] = [
  'Vanilla',
  'Pistachio',
  'Raspberry',
].map((item, index) => mockOption(index.toString(), item, index));

jest.mock('hooks/usePollOptions', () => () => mockPollOptions);

describe('<WrongMaxChoiceIndicator/>', () => {
  it('indicates when there are fewer poll answer options than maximum number of answers participant can give', () => {
    render(<WrongMaxChoiceIndicator questionId="questionId" maxAnswers={10} />);

    expect(screen.getByTestId('wrongMaxChoiceIndicator')).toBeInTheDocument();
  });

  it('does not show when there are enough (equal number or more) poll answer options than maximum number of answers participant can give', () => {
    render(<WrongMaxChoiceIndicator questionId="questionId" maxAnswers={3} />);

    expect(
      screen.queryByTestId('wrongMaxChoiceIndicator')
    ).not.toBeInTheDocument();
  });
});
