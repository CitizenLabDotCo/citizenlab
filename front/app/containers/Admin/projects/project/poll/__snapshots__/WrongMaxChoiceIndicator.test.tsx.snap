// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WrongMaxChoiceIndicator/> boundaries renders correctly when maxAnswers is null 1`] = `""`;

exports[`<WrongMaxChoiceIndicator/> boundaries renders correctly when options is Error 1`] = `""`;

exports[`<WrongMaxChoiceIndicator/> boundaries renders correctly when options is null 1`] = `""`;

exports[`<WrongMaxChoiceIndicator/> boundaries renders correctly when options is undefined 1`] = `""`;

exports[`<WrongMaxChoiceIndicator/> renders nothing when max answers is as expected renders correclty when 2 <= maxAnswers <= pollOptions.length 1`] = `""`;
